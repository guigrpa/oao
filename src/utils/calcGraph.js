// @flow

import type { AllSpecs, OaoSpecs } from './types';
import { DEP_TYPES } from './constants';

const calcGraph = (allSpecs: AllSpecs): Array<string> => {
  const out = [];
  const pkgNames = Object.keys(allSpecs);
  if (!pkgNames.length) return out;

  // Build virtual root node
  const virtualRootDeps = {};
  pkgNames.forEach(name => {
    virtualRootDeps[name] = true;
  });
  const virtualRootNode: any = {
    name: '__VIRTUAL_ROOT__',
    specs: { dependencies: virtualRootDeps },
  };

  // Build graph starting from virtual root node, then remove it
  buildGraph(allSpecs, virtualRootNode, pkgNames, out);
  return out.slice(0, out.length - 1);
};

export const calcGraphAndReturnAsAllSpecs = (allSpecs: AllSpecs): AllSpecs => {
  const newAllSpecs = {};
  const orderedPackages = calcGraph(allSpecs);
  orderedPackages.forEach(pkg => {
    newAllSpecs[pkg] = allSpecs[pkg];
  });
  return newAllSpecs;
};

const buildGraph = (
  allSpecs: AllSpecs,
  pkg: OaoSpecs,
  pkgNames: Array<string>,
  out: Array<string>,
  visited?: Array<string> = []
) => {
  const { name } = pkg;
  visited.push(name);
  const internalDeps = getInternalDeps(pkg, pkgNames);
  for (let i = 0; i < internalDeps.length; i++) {
    const depName = internalDeps[i];
    if (visited.indexOf(depName) >= 0) continue;
    buildGraph(allSpecs, allSpecs[depName], pkgNames, out, visited);
  }
  out.push(name);
};

const getInternalDeps = (pkg: OaoSpecs, pkgNames: Array<string>) => {
  const { specs } = pkg;
  const internalDeps = {};
  for (let i = 0; i < DEP_TYPES.length; i++) {
    const depType = DEP_TYPES[i];
    const deps = specs[depType] || {};
    const depNames = Object.keys(deps);
    for (let k = 0; k < depNames.length; k++) {
      const pkgName = depNames[k];
      if (pkgNames.indexOf(pkgName) >= 0) internalDeps[pkgName] = true;
    }
  }
  return Object.keys(internalDeps);
};

export default calcGraph;
