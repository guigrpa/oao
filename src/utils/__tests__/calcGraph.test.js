/* eslint-env jest */

import calcGraph, { calcGraphAndReturnAsAllSpecs } from '../calcGraph';

const ALL_SPECS_NO_CYCLE = {
  a: {
    name: 'a',
    specs: {
      dependencies: { b: '*', c: '*' },
    },
  },
  b: {
    name: 'b',
    specs: {
      dependencies: { ext: '*' },
    },
  },
  c: {
    name: 'c',
    specs: {
      dependencies: { ext: '*' },
      devDependencies: { d: '*' },
    },
  },
  d: {
    name: 'd',
    specs: {},
  },
};

const ALL_SPECS_CYCLE = {
  a: {
    name: 'a',
    specs: {
      dependencies: { b: '*', c: '*' },
    },
  },
  b: {
    name: 'b',
    specs: {
      dependencies: { ext: '*' },
    },
  },
  c: {
    name: 'c',
    specs: {
      dependencies: { ext: '*' },
      devDependencies: { d: '*', e: '*' },
    },
  },
  d: {
    name: 'd',
    specs: {},
  },
  e: {
    name: 'e',
    specs: {
      optionalDependencies: { a: '*' },
    },
  },
};

const ALL_SPECS_MULTI_ROOT = {
  a: { name: 'a', specs: {} },
  b: { name: 'b', specs: {} },
  c: { name: 'c', specs: {} },
};

describe('buildGraph', () => {
  it('calculates a directed acyclic graph correctly when no cycles are present', () => {
    const dag = calcGraph(ALL_SPECS_NO_CYCLE);
    expect(dag).toEqual(['b', 'd', 'c', 'a']);
  });

  it('handles multiple roots correctly', () => {
    const dag = calcGraph(ALL_SPECS_MULTI_ROOT);
    expect(dag).toEqual(['a', 'b', 'c']);
  });

  it('handles cycles correctly', () => {
    const dag = calcGraph(ALL_SPECS_CYCLE);
    expect(dag).toEqual(['b', 'd', 'e', 'c', 'a']);
  });

  it('returns list as an allSpecs object', () => {
    const allSpecs = calcGraphAndReturnAsAllSpecs(ALL_SPECS_NO_CYCLE);
    expect(allSpecs).toMatchSnapshot();
  });
});
