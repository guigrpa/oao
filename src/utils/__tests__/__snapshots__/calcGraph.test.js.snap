// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildGraph returns list as an allSpecs object 1`] = `
Object {
  "a": Object {
    "name": "a",
    "specs": Object {
      "dependencies": Object {
        "b": "*",
        "c": "*",
      },
    },
  },
  "b": Object {
    "name": "b",
    "specs": Object {
      "dependencies": Object {
        "ext": "*",
      },
    },
  },
  "c": Object {
    "name": "c",
    "specs": Object {
      "dependencies": Object {
        "ext": "*",
      },
      "devDependencies": Object {
        "d": "*",
      },
    },
  },
  "d": Object {
    "name": "d",
    "specs": Object {},
  },
}
`;
